iree/turbine/kernel/wave/constraints.py

```diff
@@ -59,6 +59,58 @@ class MMAType(Enum):
     I32_32x32x16_I8 = 0x12C1
 
 
+@dataclass
+class GenericDot:
+    """
+    mma implemented through vector dot products intead of hw intrinsics.
+
+    `out_vec_size`: size of the output matrix vector
+    `k_vec_size`: size of the reduction dimension vector
+    `k_mult`: number of reduction dimension vectors
+    """
+
+    out_vec_size: int = 1
+    k_vec_size: int = 4
+    k_mult: int = 1
+
+    def get_shape(self, threads_per_wave: int) -> tuple[int, int, int]:
+        return (
+            self.out_vec_size,
+            threads_per_wave // self.k_mult,
+            self.k_vec_size * self.k_mult,
+        )  # M x N x K
+
+    def get_index_offset(
+        self, lane: IndexExpr, threads_per_wave: int
+    ) -> tuple[IndexExpr, IndexExpr, IndexExpr]:
+        return [
+            Piecewise((lane % self.out_vec_size, ~MMA_ACC), (0, MMA_ACC)),  # M
```

Comment:
Distributing across M or N dimension is completely symmetrical, but I only Implemented N for now.

Reply:
makes sense.